### PR TITLE
Track unrealized justification/finalization in forkchoice

### PR DESCRIPTION
**Pull request by potuz**
Tracks unrealized justification and finalization in forkchoice. It does not introduce runtime changes. Pure functions with their respective unit tests.

### CODE REVIEW

**Review by rauljordan — COMMENTED** [review]
After trying to grok some of the unit tests here, I have a concern about the function signature of:
```go
func (f *ForkChoice) InsertOptimisticBlock(
	ctx context.Context,
	slot types.Slot,
	blockRoot, parentRoot, payloadHash [fieldparams.RootLength]byte,
	justifiedEpoch, finalizedEpoch types.Epoch,
) 
```
It feels very likely someone will mix up the last two function arguments, and even the first three. Highly recommend we switch to a struct of func arguments as an alternative. It will also make unit tests more understandable

File: beacon-chain/forkchoice/doubly-linked-tree/store.go [hunk 1]
```diff
@@ -145,8 +147,12 @@ func (s *Store) insert(ctx context.Context,
...
 // updateCheckpoints Update the justified / finalized epochs in store if necessary.
 func (s *Store) updateCheckpoints(justifiedEpoch, finalizedEpoch types.Epoch) {
-	s.justifiedEpoch = justifiedEpoch
-	s.finalizedEpoch = finalizedEpoch
+	if s.justifiedEpoch < justifiedEpoch {
+		s.justifiedEpoch = justifiedEpoch
+	}
+	if s.finalizedEpoch < finalizedEpoch {
+		s.finalizedEpoch = finalizedEpoch
+	}
```
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/store.go:155]: Do you want to log some warnings if someone attempts to update justified/finalized epochs less than current?

File: beacon-chain/forkchoice/doubly-linked-tree/unrealized_justification.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	return nil
+}
+
+// UpdateUnrealizedCheckpoints "realizes" the unrealized justified and finalized
+// epochs stored within nodes. It should be called at the beginning of each
+// epoch
+func (f *ForkChoice) UpdateUnrealizedCheckpoints() {
+	f.store.nodesLock.Lock()
+	defer f.store.nodesLock.Unlock()
+	for _, node := range f.store.nodeByRoot {
```
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/unrealized_justification.go:41]: Maybe a silly question, do we need to do this for all the nodes? or only the tips?

File: beacon-chain/forkchoice/doubly-linked-tree/unrealized_justification_test.go [hunk 3]
```diff
@@ -0,0 +1,186 @@
+package doublylinkedtree
+
+import (
+	"testing"
+
+	"github.com/prysmaticlabs/prysm/config/params"
+	types "github.com/prysmaticlabs/prysm/consensus-types/primitives"
+	"github.com/prysmaticlabs/prysm/testing/require"
+	"golang.org/x/net/context"
```
Comment (terencechain) [inline beacon-chain/forkchoice/doubly-linked-tree/unrealized_justification_test.go:9]: why use a different library? 

```suggestion
	"context"
```

File: beacon-chain/forkchoice/protoarray/unrealized_justification.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+	s.nodesLock.Lock()
+	defer s.nodesLock.Unlock()
+	index, ok := s.nodesIndices[root]
+	if !ok {
+		return ErrUnknownNodeRoot
+	}
+	if index >= uint64(len(s.nodes)) {
+		return errInvalidNodeIndex
+	}
+	node := s.nodes[index]
```
Comment (terencechain) [inline beacon-chain/forkchoice/protoarray/unrealized_justification.go:17]: Check ok too?

```suggestion
	node, ok := s.nodes[index]
```

**Review by terencechain — APPROVED** [review]

File: beacon-chain/forkchoice/protoarray/unrealized_justification.go (hunk 4 above)
Comment (potuz) [reply beacon-chain/forkchoice/protoarray/unrealized_justification.go:17]: It's impossible for that to happen given the previous check.

File: beacon-chain/forkchoice/doubly-linked-tree/unrealized_justification_test.go (hunk 3 above)
Comment (potuz) [reply beacon-chain/forkchoice/doubly-linked-tree/unrealized_justification_test.go:9]: I have no clue, this is vim-go, thanks for catching this

File: beacon-chain/forkchoice/doubly-linked-tree/unrealized_justification.go (hunk 2 above)
Comment (potuz) [reply beacon-chain/forkchoice/doubly-linked-tree/unrealized_justification.go:41]: We need only the tips but the cost of finding them involves a full loop anyway so it's better to just update then all.

**Pull request merged**